pkg/auth/oauth/registry/tokentimeouts.go

```diff
@@ -21,33 +21,41 @@ var ErrTimedout = errors.New("Token timed out")
 
 const timeoutsInPatch = `[{"op": "test", "path": "/timeoutsIn", "value": %d}, {"op": "replace", "path": "/timeoutsIn", "value": %d}]`
 
+var _ btree.Item = &tokenData{}
+
 type tokenData struct {
 	token *oauth.OAuthAccessToken
 	seen  time.Time
 }
 
 func (a *tokenData) timeout() time.Time {
-	return a.token.CreationTimestamp.Time.Add(time.Duration(a.token.TimeoutsIn) * time.Second)
+	return a.token.CreationTimestamp.Add(time.Duration(a.token.TimeoutsIn) * time.Second)
 }
 
-type tokenDataRef struct {
-	name    string
-	timeout time.Time
-}
+func (a *tokenData) Less(than btree.Item) bool {
+	other := than.(*tokenData)
+
+	// From the btree.Item docs:
+	// > If !a.Less(b) && !b.Less(a), we treat this to mean a == b (i.e. we can only hold one of either a or b in the tree)
+	// Thus we use this to guarantee that the btree will not contain duplicate entries for the same token
+	if a.token.Name == other.token.Name {
```

Comment:
I think this is broken.
Add a test where you add 6 tokens t1, t2, t3, t4, t5, t6 in this order
Where t1 = {"A", 1}, t2 = {"B", 2}, t3 = {"A", 3}, t4 = {"B", 4}, t5 = {"A", 5}, t6 = {"B", 6}

See if you get what you expect.
The reason I sued a map is that it makes it quick to find and remove the duplicate tokens, the only way to do that in a generic way with a btree is to go through the whole tree every time. Unless you can make specific assumptions about timeouts (Ie assume tokens always have the same timeout for the same name, but I do not think you can make that assumption in a multimaster environment)

Reply:
I already have a test for this, but I added this scenario as well (and it works as expected).

> the only way to do that in a generic way with a btree is to go through the whole tree every time

That is literally what `InsertOrReplace` does (using a binary search based on `Less`).

> but I do not think you can make that assumption in a multimaster environment

I do not understand what multiple masters has to do with this.

Reply:
The test passes because the elemnts are initialized wrongly.
You set seen but not CrateTimestamp, which is the one we use for oredring in Less(). So in your test all tokens have the same time+timeout and they all order to the bottom, so they all are always relying exclusively on name ordering.

If you properly add the CreateTimestamp then your test fails as I predicted.

Reply:
See it here:
https://github.com/simo5/origin/commit/b350f59b376d9ff045bd0e1fa036b560ceed12ab